cis194/week4/daniel-deng/src/Exercise01.hs

```diff
@@ -0,0 +1,13 @@
+module CIS194.Homework04.Exercise01 (fun1, fun2) where
+
+import Data.Bool
+
+fun1 ::  [Integer] -> Integer
+fun1 = product . fmap (subtract 2) . filter even
+
+fun2 :: Integer -> Integer
+fun2 = bool <$> handleN <*> const 0 <*> (<=1)
+     where
+       handleN    = bool <$> handleOdd <*> handleEven <*> even
+       handleOdd  = fun2 . (1+) . (3*)
+       handleEven = (+) <*> fun2 . (`div` 2)
```

Comment:
> `handleOdd` and `handleEven` are recursive on `fun2`. Can use utilise `iterate` instead?

@shineli1984 yes we certainly can. My question is, would we prefer wholemeal programming in general and try to avoid recursive function calls? In the production codebase, for this particular instance, do we prefer to have `iterate` and `takeWhile`?

Reply:
They are both fine in terms of `wholemeal`. But most recursions (as in recursion with a lot of operational concerns around it) are not.